ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3MultipartUploadPartHandler.java

```diff
@@ -65,28 +107,202 @@ void handle(RestRequest restRequest, RestResponseChannel restResponseChannel,
     // TODO [S3] : set part ttl value
     // restRequest.setArg(TTL, xxx);
 
-    // 3. Intercept the response callback and make needed changes in response headers.
-    Callback<Void> wrappedCallback = (result, exception) -> {
-      if (exception != null) {
-        callback.onCompletion(null, exception);
-        return;
-      }
+    restRequest.setArg(SEND_FAILURE_REASON, Boolean.TRUE);
+
+    // TODO [S3] verifyChunkAccountAndContainer? getAndVerifyReservedMetadataBlobId?
+    new S3MultipartUploadPartHandler.CallbackChain(restRequest, restResponseChannel, callback).start();
+  }
+
+  /**
+   * Represents the chain of actions to take. Keeps request context that is relevant to all callback stages.
+   */
+  private class CallbackChain {
+    private final RestRequest restRequest;
+    private final RestResponseChannel restResponseChannel;
+    private final Callback<ReadableStreamChannel> finalCallback;
+    private final String uri;
+
+    /**
+     * @param restRequest the {@link RestRequest}.
+     * @param restResponseChannel the {@link RestResponseChannel}.
+     * @param finalCallback the {@link Callback} to call on completion.
+     */
+    private CallbackChain(RestRequest restRequest, RestResponseChannel restResponseChannel,
+        Callback<ReadableStreamChannel> finalCallback) {
+      this.restRequest = restRequest;
+      this.restResponseChannel = restResponseChannel;
+      this.finalCallback = finalCallback;
+      this.uri = restRequest.getUri();
+    }
+
+    /**
+     * Start the chain by calling {@link SecurityService#processRequest}.
+     */
+    private void start() {
+      restRequest.getMetricsTracker()
+          .injectMetrics(frontendMetrics.putBlobMetricsGroup.getRestRequestMetrics(restRequest.isSslUsed(), false));
       try {
```

Comment:
Any exception should be caught in S3BaseHandler, shouldn't need to catch here.

Reply:
You are right. Removed.